File: src/main/java/Q53_MaximumSubarray.java
/***
 * 1hr
 * Runtime: 0 ms, faster than 100.00% of Java online submissions for Maximum Subarray.
 * Memory Usage: 39.4 MB, less than 66.18% of Java online submissions for Maximum Subarray.
 */
public class Q53_MaximumSubarray {
    public int maxSubArray(int[] nums) {
        if(nums == null || nums.length == 0) {
            return 0;
        }
        if(nums.length == 1) {
            return nums[0];
        }
        int min = nums[0];
        int max = nums[0];
        for(int i = 1; i < nums.length; i++) {
            nums[i] = nums[i] + nums[i - 1];
            if(nums[i] > max) {
                max = nums[i];
            }
            if(nums[i] < min) {
                min = nums[i];
            }
        }
        return max - min;
    }
}

File: LeetCodePractice/LeetCodePractice/Q53_MaximumSubarray.cpp
#include "LCHeader.h"
class Solution {//4:15 19
public:
    int maxSubArray(vector<int>& nums) {
        int max = INT_MIN;
        for (auto& x : nums) {
            if (x > max) {
                max = x;
            }
        }
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] < 0) {
                continue;
            }
            int sum = 0;
            while (i < nums.size() && sum + nums[i] >= 0) {
                sum += nums[i];
                i++;
                max = std::max(max, sum);
            }
        }
        return max;
    }
};